### PR TITLE
Create interactive repo when creating a user

### PR DESCRIPTION
**Pull request by ghickman**
This updates the Create interactive user page to also create a repo on GitHub if it doesn't already exist.

I was struggling to find nice patterns to create methods on Project/User which match those in `interactive/commands.py`.  I felt like the wrong layer to have a Model know about an API, so in the end I tried pulling all the functionality into "commands", as per [CQRS terminology](https://martinfowler.com/bliki/CommandQuerySeparation.html).  I'm not sold on following CQRS strictly but I've been [curious to see how useful it would be in job-server for a while](https://github.com/opensafely-core/job-server/issues/827) and this feels like a reasonable place to give it a go, being fairly separate from the rest of the system, and tasks which require more than one Model or API call.

Fixes: #2617

### CODE REVIEW

File: jobserver/github.py
Comment (bloodearnest) [inline jobserver/github.py:193]: Hmm, so I'm not sure about using the research-template for interactive, as it's setup with base files we don't want, and a README that's misleading in this case. Also, it runs an initial action to render the repo template, which given we're automating things could be a bit messy.

I think it might be better to create a blank repo, and include a standard boiler plate README.md in every commit that spells out that this was a repo for an Interactive Project, with links to the project/workspace on jobs.opensafely.org.

File: interactive/commands.py
Comment (lucyb) [inline interactive/commands.py:6]: I think this is a nice way of moving logic out of the view and model, and has resulted in some cleaner code.

My only, minor, comment is that I'm always a bit wary of naming something after the particular pattern that it implements. There's usually something that expresses the intent better within that specific domain. I'm not sure what this is right now though... we're creating jobserver things, but for Interactive. Come to think of it, all of the imports are from jobserver and this code is called from jobserver too. Is it in the right place?

Things to think about, but no reason to not merge it.

**Review by lucyb — APPROVED** [review]

File: jobserver/github.py
Comment (bloodearnest) [inline jobserver/github.py:475]: I think this implementation may wipe any existing topics that have been manually added? 

Whilst likely not a problem in the interactive case, I see a foot gun waiting to happen. e.g. we could accidentally remove our manually created `github-releases` topics via this API call.

File: interactive/commands.py
Comment (bloodearnest) [inline interactive/commands.py:21]: I've created an [Interactive](https://github.com/orgs/opensafely/teams/interactive) team as a way to group interactive repos with `write` permission?. Do you think we could add this newly created repo to that team?

I think topics are informative and useful for searching. But adding them to a team means we can control privileges better.

e.g. we could give the interactive bot account write access to just the repos on this team.

File: jobserver/github.py
Comment (bloodearnest) [reply jobserver/github.py:193]: Alternatively, maybe an alternate interactive specific template might be better? And you can pass the template in as an arg to create_repo?

**Review by bloodearnest — COMMENTED** [review]
Looks sensible to me

If I'm right about `set_repo_topics()`, I think we need to fix that.

I think the repo template question is worth figuring out too.

File: tests/verification/test_github.py
Comment (ghickman) [reply tests/verification/test_github.py:263]: Verification tests exist to verify that a Fake responds like the real thing, which is why we only run these tests in CI by default!

File: jobserver/github.py
Comment (ghickman) [reply jobserver/github.py:194]: Oh this is a good point 🤔

So the API module methods are supposed to represent what the API does, not what we do with them, following along with the idea that these modules are thin wrappers of endpoints.  With that in mind, and your point that I've hidden that this is creating a repo from a template… that should be expressed in the name! (It's even hitting a different endpoint to the normal repo one)

File: interactive/commands.py
Comment (ghickman) [reply interactive/commands.py:6]: I see your point about the code all coming from job-server, and I did think about putting this module into the `jobserver` app, but they're so specific to interactive that they would all need renaming to denote that.

My long term hope/goal is that we can move all onboarding to a flow similar to this, at which point I'd move them into the main app because they're not special-cased, but until then they're specifically for an interactive flow.

File: jobserver/github.py
Comment (ghickman) [reply jobserver/github.py:475]: Good point.  I'll get the current topics and merge in anything we pass to avoid the footgun 👍  (You can't do add/remove via the API, you have to do GET/PUT)

File: interactive/commands.py
Comment (ghickman) [reply interactive/commands.py:21]: Absolutely lets add it to the team 👍

File: jobserver/github.py
Comment (ghickman) [reply jobserver/github.py:193]: The current research-template's README is already a boilerplate!  I'm not sure we need anything in there about interactive?

A separate template might well work but I'm leaning towards not doing that. We're going to wipe the working directory each time we generate code anyway.  Also it saves maintaining another repo!

I've added the ability to pass the template org/repo and renamed the method slightly to make it clearer what we're doing under the hood.

Comment (bloodearnest) [reply jobserver/github.py:193]: Yes, but it's boilerplate for a very different thing - a manually authored study.

I think the interactive repos should have a README something like:

```
# OpenSAFELY Interactive Repository for Project Foo

This repo contains the code for the [OpenSAFELY Interactive] study for [project Foo].  Unlike regular studies, this code is not hand written, and is instead automatically generated from [a template], 

Any published results from this study will be available here: ...
```

This could possibly be rendered on every commit we do, and is not integral to walking skeleton, so we can defer decision.

But my other concern I think make cause issues in tests.

When you clone a repo from the research-template, it runs an initial action that updates the README template and then deletes itself. This takes a minute or two to run.

It's entire possible, especially in automated testing, that we a) create repo b) clone c) initial setup job finishes, making our clone out of date b) we commit/push, and thus error.

File: jobserver/github.py
Comment (bloodearnest) [reply jobserver/github.py:475]: yep, familiar with this pain :(

Comment (ghickman) [reply jobserver/github.py:502]: topics are part of the get_repo response so I can check them there 🎉  Updated the calls in the create_repo command to make sure we use the existing ones and only update if it's not already there.

File: jobserver/github.py
Comment (ghickman) [reply jobserver/github.py:220]: Adding it to the analysis template repo for now sounds fine but I'm keen to only have one readme template if possible, that was a big point of the work when changing it for repo sign-offs!  I have a feeling we can handle the differences on job-server rather than in repo READMEs _but_ we should discuss this with the OSI team too (I'll drop a note in channel).

Comment (ghickman) [reply jobserver/github.py:220]: Wouldn't making it part of the template repo mean it was the same for all actions…?  My goal was to have it all the same so now I'm a bit confused why having it in the template directory with nothing changed would not achieve that…!

Comment (bloodearnest) [reply jobserver/github.py:220]: Not as currently implemented no, as each templated analysis is separate.

Additionally, we'd need add additional data in the template context.

I think this belongs in InteractiveAnalysisTemplate - its a step analogous to downloading a codelist, the same process for rendering all templated analysis.

Comment (ghickman) [reply jobserver/github.py:217]: Switched to creating a repo without the research template as discussed 👍

**Pull request merged**
